src/net.h

```diff
@@ -417,6 +419,179 @@ class V1Transport final : public Transport
     size_t GetSendMemoryUsage() const noexcept override EXCLUSIVE_LOCKS_REQUIRED(!m_send_mutex);
 };
 
+class V2Transport final : public Transport
+{
+private:
+    /** Contents of the version packet to send. BIP324 stipulates that senders should leave this
+     *  empty, and receivers should ignore it. Future extensions can change what is sent as long as
+     *  an empty version packet contents is interpreted as no extensions supported. */
+    static constexpr std::array<std::byte, 0> VERSION_CONTENTS = {};
+
+    // The sender side and receiver side of V2Transport are state machines that are transitioned
+    // through, based on what has been received. The receive state corresponds to the contents of,
+    // and bytes received to, the receive buffer. The send state controls what can be appended to
+    // the send buffer.
+
+    /** State type that defines the current contents of the receive buffer and/or how the next
+     *  received bytes added to it will be interpreted.
+     *
+     * Diagram:
+     *
+     *   start                                            /---------\
+     *     |                                              |         |
+     *     v                                              v         |
+     *    KEY -> GARB_GARBTERM -> GARBAUTH -> VERSION -> APP -> APP_READY
+     */
+    enum class RecvState : uint8_t {
+        /** Public key.
+         *
+         * This is the initial state, during which the other side's public key is
+         * received. When that information arrives, the ciphers get initialized and the state
+         * becomes GARB_GARBTERM. */
+        KEY,
+
+        /** Garbage and garbage terminator.
+         *
+         * Whenever a byte is received, the last 16 bytes are compared with the expected garbage
```

Comment:
5cb433948b050913d22914905c7db510e9ef48d2
Should be extra careful with not keeping 4kb in memory waiting here i guess... After staring at the code for a few minutes I struggle to find this logic :) So perhaps worth adding a comment for it?

This question possibly applies to the whole transport thing. Or maybe I'm just confused with how it goes through spans in the first place...

Reply:
Later communication is subject to ping timeout, e.g. if the peer starts sending us a new block, sends 1MB and stops any sends to us, the peer will eventually be disconnected due to ping timeout. I guess this handshake should be subject to the same. Is it? I guess that is in some later PR which will engage the V2 transport.

Reply:
>Regarding the 4 kB, it's wiped in ProcessReceivedGarbageBytes now: when the full garbage + terminator has been processed, m_recv_buffer.clear(); is called.

What will happen if we're at 3.99 Kb of garbage and no terminator? Am i right that the caller would be responsible for terminating this after a timeout?